### PR TITLE
Git - respect configured submodule names in SCM

### PR DESCRIPTION
**Pull request by Gravifer**
## Summary

- respect customized submodule names from the superproject's `.gitmodules` throughout SCM UI
- add a creation-time repository name to the proposed `scmProviderOptions` API
- preserve existing workspace-folder and basename behavior for ordinary repositories, worktrees, generated submodule names, and independently opened submodules
- use the same resolved name in the built-in Git repository picker

Addresses #269812

## Behavior

A submodule discovered through an open superproject uses its configured name only when that name differs from its path. The explicit name flows into `ISCMProvider.name`, so the repositories tree, active repository status, SCM and history pickers, accessibility text, and Git repository picker remain consistent. Opening the submodule independently retains the current behavior.

## Validation

- `npm run typecheck-client`
- `npm run gulp compile-extension:git`
- `npm run valid-layers-check`
- `git diff --check upstream/main...HEAD`

The full `compile-extensions` command was also attempted; it stopped in the unrelated CSS language extension because that extension's separate Node typings were not installed in this clone. Focused core and Git test launchers could not start because the local `.build/electron` test runtime is absent.

### CODE REVIEW

**Comment by Gravifer** [comment]
@microsoft-github-policy-service agree

**Comment by Gravifer** [comment]
Current attempts at PR head heavily change method signatures; I hope to reduce the footprint eventually.

**Comment by Gravifer** [comment]
OK, I think I can at least have Copilot go through my jank now. Sorry for LLM generated commit messages btw, I may squash everything later.

File: extensions/git/src/model.ts [hunk 1]
```diff
@@ -199,7 +204,7 @@ export class Model implements IRepositoryResolver, IBranchProtectionProviderRegi
 	private openRepositories: OpenRepository[] = [];
 	get repositories(): Repository[] { return this.openRepositories.map(r => r.repository); }
 
-	private possibleGitRepositoryPaths = new Set<string>();
+	private possibleGitRepositoryPaths = new Map<string, string | undefined>(); // a set with optional labels
```
[the remark this answers is not in the document]
Comment (Gravifer) [reply extensions/git/src/model.ts]: I don't even think this should be left as a Map... I hope there to be input from someone (that at least has ever written TS, as opposed to myself) to suggest a best practice instead of this.

File: extensions/git/src/model.ts [hunk 2]
```diff
@@ -23,7 +23,12 @@ import { ISourceControlHistoryItemDetailsProviderRegistry } from './historyItemD
 
 class RepositoryPick implements QuickPickItem {
 	@memoize get label(): string {
-		return path.basename(this.repository.root);
+		const defaultLabel = path.basename(this.repository.root);
+		const suppliedLabel = this.repository.repoLabel;
+		if (!suppliedLabel || path.basename(suppliedLabel) === defaultLabel) {
```
Comment (Gravifer) [inline extensions/git/src/model.ts]: This is janky! 

- [ ] Is there an `or_else` in TS?
- [ ] Is the [logic](https://github.com/microsoft/vscode/issues/269812#issuecomment-3369016426) I outlined the way to go?

**Comment by Gravifer** [comment]
I have substantially revised the approach taken.

My original attempt carried an optional label through the repository scan queue, `Model.openRepository`, `Git.open`, and the low-level Git repository object. That label was then used only by the Git extension’s repository quick picker. Besides changing several unrelated method signatures, this produced an incomplete result: the picker could show the configured submodule name while the persistent SCM tree, status bar, history views, accessibility text, and other SCM surfaces continued to show the directory name. The result could also depend on repository discovery order.

The branch tip removes that plumbing. Instead, it adds an optional, creation-time repository name to the existing proposed `scmProviderOptions` API. The name is transported through the extension host and assigned to the core `ISCMProvider.name`. This is the value the workbench already uses throughout its SCM user interface, so the change applies consistently without modifying each renderer or view individually.

The Git extension now determines the name at the point where it creates the child repository’s source control. If the child is a submodule, its superproject is open, and the matching `.gitmodules` entry has a logical name different from its path, that configured name is supplied to SCM core. Git’s own repository picker uses the same resolved value.

This yields deliberately contextual behavior:

- A submodule discovered through its superproject uses the customized `.gitmodules` name.
- A submodule created while its matching superproject is not open retains the existing workspace-folder or directory name.
- A normal, automatically generated submodule name—where `name === path`—does not override existing behavior.
- Ordinary repositories and worktrees are unaffected.

I think this distinction is useful. When a repository is opened independently, its folder or workspace name is the most immediate identity available to the user. When it is discovered as part of a superproject, however, the superproject’s `.gitmodules` file supplies an intentional logical identity. That identity can be much more informative than the final path segment, particularly for deeply nested submodules or layouts where the directory name is generic. The logic is also consistent across the repository tree, status and history surfaces, SCM pickers, accessibility output, and the Git picker.

A customized submodule name takes precedence over a workspace-folder alias when the child is created while its superproject is open and resolvable. This avoids returning to the earlier split behavior, where different SCM surfaces could disagree about the same repository. Existing SCM rendering already handles duplicate names by showing parent and path context where necessary, so no new disambiguation mechanism is required.

The implementation is intentionally narrow from a maintenance perspective. It does not alter the stable extension API, add a setting, create persistent state, introduce a migration, or require dynamic rename events. The proposed API gains one optional final argument; callers that omit it retain the current workspace-folder, basename, and provider-label fallbacks. No existing extension needs to change.

Likewise, the workbench renderers are untouched. They already consume `ISCMProvider.name`, so the new value flows naturally to the existing UI. The Git-side lookup happens once during repository creation, uses submodule information that has already been parsed, and does not start additional Git processes or watchers. The earlier changes to the scan queue and low-level Git abstraction have been removed.

I hope this would not introduce much burden for the maintainers or become obstructive for future contributors. Thanks to you all for the excellent project that benefits all developers.

**Comment by Gravifer** [comment]
Hi @lszomoru — when you have a chance, would you mind taking a look at the revised approach? The PR now routes repository display names through SCM core and removes the original Git-specific discovery plumbing. I’d particularly appreciate feedback on whether this direction fits the SCM API. No rush, and thank you for your time.

@coderabbitai @copilot can you initiate reviews as well? Perhaps focus on whether there are potential drawbacks to this.

------
I'm force pushing to make rebases from time to time. The material commits are:
- feat(scm): declare repository display names [wip] [no-ci]
- feat(scm): transport repository display names [wip] [no-ci]
- feat(scm): apply repository display names
- feat(git): select customized submodule names
- feat(git): name parent-detected submodules
- feat(git): align repository picker names
With moving SHAs.

**Comment by Gravifer** [comment]
> Quoting the @copilot updated [overview](https://github.com/microsoft/vscode/tasks/92d32eca-7f35-4695-bac5-6803a4bb3dc6?session_id=4642bb99-b2f2-493c-944e-4c18337ec9e5) here

Extends SCM naming so Git submodules display customized .gitmodules names consistently while preserving existing fallback behavior.

**Changes**:
- Adds an optional creation-time repository name to the proposed SCM API and RPC flow.
- Resolves customized submodule names for SCM providers and Git repository picks.
- Adds coverage for name resolution and forwarding.

### Reviewed Changes

File | Description
-- | --
`src/vscode-dts/vscode.proposed.scmProviderOptions.d.ts` | Adds the optional SCM repository name API.
`src/vs/workbench/api/test/common/extHostSCM.test.ts` | Tests name forwarding.
`src/vs/workbench/api/common/extHostSCM.ts` | Passes names through the extension host.
`src/vs/workbench/api/common/extHost.protocol.ts` | Extends the SCM registration protocol.
`src/vs/workbench/api/common/extHost.api.impl.ts` | Exposes and proposal-gates the new argument.
`src/vs/workbench/api/browser/mainThreadSCM.ts` | Applies explicit names with existing fallbacks.
`extensions/git/src/test/git.test.ts` | Tests customized-name resolution.
`extensions/git/src/repository.ts` | Resolves and registers submodule names.
`extensions/git/src/model.ts` | Uses resolved names in repository picks.
`extensions/git/src/git.ts` | Adds the submodule display-name helper.

File: src/vs/workbench/contrib/scm/browser/scmRepositoryRenderer.ts
Comment (Gravifer) [inline src/vs/workbench/contrib/scm/browser/scmRepositoryRenderer.ts:161]: The repository renderer displays `ISCMProvider.name`

File: src/vs/workbench/contrib/scm/browser/scmViewService.ts
Comment (Gravifer) [inline src/vs/workbench/contrib/scm/browser/scmViewService.ts:527]: These values are normally identical, so the distinction has little effect today. With this PR, though, an explicit provider name can intentionally differ from the repository path.

Comment (Gravifer) [inline src/vs/workbench/contrib/scm/browser/scmViewService.ts]: "Sort by Name," however, still derives its sort key independently from the workspace-folder name or root basename

**Review by Gravifer — COMMENTED** [review]
### Minor follow-up from local validation

For example:

| Configured name | Root basename | Displayed label |
|---|---|---|
| `zebra` | `alpha` | `parent / zebra` |
| `apple` | `zulu` | `parent / apple` |

The current implementation displays `zebra` and `apple`, but sorts using `alpha` and `zulu`. Selecting “Sort by Name” could therefore produce:

```text
parent / zebra
parent / apple
```

The PR already makes the explicit name authoritative while preserving the existing workspace-folder, basename, and provider-label fallbacks.



I am going to try a small follow-up correction that sorts by the displayed provider name:

```ts
const nameComparison = compareFileNames(
	op1.repository.provider.name,
	op2.repository.provider.name
);
```

The existing path tie-breaker and locale-aware comparator would remain unchanged.

I think this is justified within the PR because customized names would otherwise be displayed correctly but behave unexpectedly under the existing “Sort by Name” option. For callers that do not supply an explicit name, `provider.name` already resolves through the current fallbacks, so ordinary repositories should retain their existing ordering.

I do not plan to modify the pre-existing comparer tests. Their initial local failures came from the direct Electron test host inheriting `zh-CN`; all 51 relevant tests passed when the host was run with `en-US`. That appears separate from this small SCM sort-key adjustment.

**Comment by Gravifer** [comment]
I have pushed the follow-up correction discussed above.

- [`2b68636`](https://github.com/Gravifer/microsoft.vscode/commit/2b68636) synchronizes the SCM transport test with the asynchronous RPC protocol before asserting registration and after disposal.
- [`e4e1568`](https://github.com/Gravifer/microsoft.vscode/commit/e4e1568) makes “Sort by Name” use the same canonical `ISCMProvider.name` displayed by the workbench. The existing locale-aware comparator and root-path tie-breaker remain unchanged.

https://github.com/Gravifer/microsoft.vscode/blob/e4e1568/src/vs/workbench/contrib/scm/browser/scmViewService.ts#L520-L529

The new regression test deliberately opposes provider names and root basenames, while giving two repositories the same provider name in reverse path order. It therefore verifies both that the displayed name controls sorting and that the existing path tie-breaker is preserved.

https://github.com/Gravifer/microsoft.vscode/blob/e4e1568/src/vs/workbench/contrib/scm/test/browser/scmViewService.test.ts#L36-L77

The transport test now explicitly waits for the test RPC queue:

https://github.com/Gravifer/microsoft.vscode/blob/2b68636/src/vs/workbench/api/test/common/extHostSCM.test.ts#L56-L100

Local validation included:

- `typecheck-client`: passed
- Targeted ESLint: passed
- Client esbuild transpilation: passed
- Focused `SCMViewService` suite: 2 passing
- Focused `ExtHostSCM` suite: 3 passing
- Full Git integration suite: 59 passing
- Broad Electron suite: 27,898 passing, 453 pending, 11 failing

None of the 11 Electron failures involved Git or SCM, and all occurred in source/test files byte-identical to upstream `main`. Their symptoms were environment-sensitive: Windows environment-variable assumptions, loopback tests interacting with the local proxy, a localized Kerberos error string, display-scale-dependent measurement, and timeouts. I stopped rather than modifying or repeatedly retrying unrelated tests. The four comparer tests that had previously exposed the host `zh-CN` locale all passed when the Electron host was run with VS Code’s `en-US` language.

The targeted coverage command also passed its tests, but the existing Electron coverage path produced an empty `lcov.info`, so I am not claiming numerical coverage.

These results were obtained on the patch-equivalent validation worktree immediately before the branch was refreshed onto newer `main`; range-diff confirmed that the PR patches and the two follow-up commits were unchanged.

**Comment by Gravifer** [comment]
Hi @lszomoru — I'm still interested in contributing this improvement and would be grateful for your thoughts on the revised approach, particularly the optional SCM provider name API. I'm happy to adjust the implementation based on your guidance. If someone else would be better placed to offer input, I would also appreciate a pointer. Thank you for your time.

I've only stopped rebasing the branch so that we have a stable review surface; if rebasing to trunk tip is still desired, do please instruct me to do so :D

**Comment by Gravifer** [comment]
I have pushed one final branch refresh at [`53f0c69`](https://github.com/Gravifer/microsoft.vscode/commit/53f0c69), based directly on current `main` `1f625ad`. I plan to keep the PR head stable from here.

During the private hardening pass I found a discovery-order race in the revised implementation: the child could be created before the open parent's first status operation published its parsed submodule list, causing the basename to be frozen as the creation-time SCM name. The [current resolution](https://github.com/Gravifer/microsoft.vscode/blob/53f0c69/extensions/git/src/model.ts#L708-L729) reads the parent's current `.gitmodules` before child/SCM creation, while retaining the last parsed parent state only as a fallback for genuine I/O errors.

A [real-repository integration test](https://github.com/Gravifer/microsoft.vscode/blob/53f0c69/extensions/git/src/test/submodule.test.ts#L17-L116) now drains the parent's initial status, deliberately empties its published submodule state, and opens the configured nested child through the Git model. It passes at the PR head; on the same stack without the race fix it fails with `metaprogramming` instead of `mpil`.

Current local validation includes full extension compilation, core and proposed-API typechecks, targeted ESLint, the focused SCM suites, and the full Git integration suite (59 passing, 2 pre-existing pending, 0 failing). The direct-file lookup starts no Git process or watcher; a warm-cache probe measured about 0.33 ms per child, or roughly 3.3 ms at the default 10-submodule detection limit.

I have updated the PR description with the exact behavior contract and validation details. Thank you for considering the revised and now race-hardened approach.
